README.md
# Omnipay: Przelewy24 API V1

**Przelewy24 API V1 gateway for the Omnipay PHP payment processing library**

[![CI](https://github.com/mysiar/omnipay-przelewy24v1/actions/workflows/tests.yml/badge.svg)](https://github.com/mysiar/omnipay-przelewy24v1/actions)
[![codecov](https://codecov.io/gh/mysiar/omnipay-przelewy24v1/branch/main/graph/badge.svg?token=gW4QFlc4lw)](https://codecov.io/gh/mysiar/omnipay-przelewy24v1)
[![Latest Version](https://img.shields.io/github/release/mysiar/omnipay-przelewy24v1.svg)](https://github.com/mysiar/omnipay-przelewy24v1/releases)
[![Total downloads](https://img.shields.io/packagist/dt/mysiar/omnipay-przelewy24v1.svg)](https://packagist.org/packages/mysiar/omnipay-przelewy24v1)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style)](LICENSE)
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fmysiar%2Fomnipay-przelewy24v1.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2Fmysiar%2Fomnipay-przelewy24v1?ref=badge_shield)
[![PHP Version Require](http://poser.pugx.org/mysiar/omnipay-przelewy24v1/require/php)](https://packagist.org/packages/mysiar/omnipay-przelewy24v1)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/mysiar/omnipay-przelewy24v1/badges/quality-score.png?b=main)](https://scrutinizer-ci.com/g/mysiar/omnipay-przelewy24v1/?branch=main)

## API endpoints implemented

| API endpoint                        | Gateway method   |
|:------------------------------------|:-----------------|
| /api/v1/testAccess                  | testAccess       |
| /api/v1/payment/methods             | methods          |
| /api/v1/transaction/register        | purchase         |
| /api/v1/transaction/verify          | completePurchase |
| /api/v1/transaction/refund          | refund           |
| /api/v1/transaction/by/sessionId    | purchaseInfo     |
| /api/v1/card/info                   | cardInfo         |
| /api/v1/card/pay                    | cardPay          |
| /api/v1/card/charge                 | cardCharge       |
| /api/v1/card/chargeWith3ds          | cardCharge3ds    |
| /api/v1/transaction/registerOffline | purchaseOffline  |

## Install

This gateway can be installed with [Composer](https://getcomposer.org/):

``` bash
$ composer require mysiar/omnipay-przelewy24v1
```

## Usage

The following gateways are provided by this package:

* Przelewy24

Reference official documentation https://developers.przelewy24.pl/index.php?en

## Example

```php

require_once  __DIR__ . '/vendor/autoload.php';

use Omnipay\Omnipay;

/** @var \Omnipay\Przelewy24\Gateway $gateway */
$gateway = Omnipay::create('Przelewy24');

$gateway->initialize([
    'merchantId' => 'YOUR MERCHANT ID HERE',
    'posId'      => 'YOUR POS ID HERE',
    'crc'        => 'YOUR CRC KEY HERE',
    'reportKey'  => 'YOUR REPORT KEY HERE'
    'testMode'   => true,
]);

$params = [
    'sessionId' => 2327398739,
    'amount' => 12.34,
    'currency' => 'PLN',
    'description' => 'Payment test',
    'email' => 'franek@dolas.com',
    'country' => 'PL',
    'returnUrl' => 'www.your-domain.pl/return_here',
    'notifyUrl' => 'www.your-domain.pl/notify_here',
];

$response = $gateway->purchase($params)->send();
```

For more examples check 
* [tests-api/GatewayTest.php](tests-api/GatewayTest.php)
* [tests/Message/](tests/Message/)

Optionally you can specify the payment channels.

```php
$gateway->initialize([
    //[...]
    'channel' => Gateway::P24_CHANNEL_CC,
]);

// or
$gateway->setChannel(Gateway::P24_CHANNEL_CC); 
```

Optionally you can specify language (default: en).

```php
$gateway->initialize([
    //[...]
    'language' => 'pl',
]);

// or
$gateway->setLanguage('pl'); 
```


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [0.7.0](https://github.com/mysiar/omnipay-przelewy24v1/compare/0.6.0...0.7.0)

> 4 August 2022

- Blik request handles for Przelewy24 [`#32`](https://github.com/mysiar/omnipay-przelewy24v1/pull/32)
- test-gh-self-hosted-action-runner [`#26`](https://github.com/mysiar/omnipay-przelewy24v1/pull/26)
- self hosted runners & tests sequence for PHP-7.3 [`99ab91a`](https://github.com/mysiar/omnipay-przelewy24v1/commit/99ab91ab88ccd83e32371d4804d60e6fadff5d73)
- refactor gateway methods to array $options [`f26f759`](https://github.com/mysiar/omnipay-przelewy24v1/commit/f26f759e3b5070512919f4fc583a9a9a3fff72cc)

#### [0.6.0](https://github.com/mysiar/omnipay-przelewy24v1/compare/0.5.1...0.6.0)

> 23 May 2022

- refactor-params-to-array: refactor all gateway methods to array $options [`#29`](https://github.com/mysiar/omnipay-przelewy24v1/pull/29)
- refactor gateway methods to array $options [`d18e804`](https://github.com/mysiar/omnipay-przelewy24v1/commit/d18e80421ff8bac421846eb9a3fa1370aeeb1e51)

#### [0.5.1](https://github.com/mysiar/omnipay-przelewy24v1/compare/0.5.0...0.5.1)

> 22 May 2022

- additional-clean-up (#25) [`#27`](https://github.com/mysiar/omnipay-przelewy24v1/pull/27)
- additional-clean-up [`#25`](https://github.com/mysiar/omnipay-przelewy24v1/pull/25)
- cardCharge3ds & purchaseOffline [`d43d606`](https://github.com/mysiar/omnipay-przelewy24v1/commit/d43d6068dd6f7645dbaef667bc33d19dd8c08702)

#### [0.5.0](https://github.com/mysiar/omnipay-przelewy24v1/compare/0.4.0...0.5.0)

> 21 May 2022

- [/api/v1/transaction/registerOffline] [`#23`](https://github.com/mysiar/omnipay-przelewy24v1/pull/23)
- [/api/v1/card/chargeWith3ds] [`#22`](https://github.com/mysiar/omnipay-przelewy24v1/pull/22)
- additional-tests [`#21`](https://github.com/mysiar/omnipay-przelewy24v1/pull/21)
- cardPay & cardCharge [`#20`](https://github.com/mysiar/omnipay-przelewy24v1/pull/20)
- cardCharge3ds & purchaseOffline [`a1fe2d8`](https://github.com/mysiar/omnipay-przelewy24v1/commit/a1fe2d8694077bc88a4c7aa4cb7f4e4ce626c519)
- additional-tests: Methods tests [`ad009c1`](https://github.com/mysiar/omnipay-przelewy24v1/commit/ad009c18ef068f67bd5df206fbe0b67151fe32df)
- additional-tests: TestAccess tests [`fb17f0a`](https://github.com/mysiar/omnipay-przelewy24v1/commit/fb17f0abba02b3620a9b3abc2a7ea9e8df944335)

#### [0.4.0](https://github.com/mysiar/omnipay-przelewy24v1/compare/0.3.2...0.4.0)

> 20 May 2022

- cardPay & cardCharge [`#20`](https://github.com/mysiar/omnipay-przelewy24v1/pull/20)
- [/api/v1/card/charge] [`#19`](https://github.com/mysiar/omnipay-przelewy24v1/pull/19)
- [/api/v1/card/pay] [`#18`](https://github.com/mysiar/omnipay-przelewy24v1/pull/18)
- main: update [`72254b8`](https://github.com/mysiar/omnipay-przelewy24v1/commit/72254b871811f6580550f0ab7a941ee3eaa32eb3)

#### [0.3.2](https://github.com/mysiar/omnipay-przelewy24v1/compare/0.3.1...0.3.2)

> 19 May 2022

- fixes [`#17`](https://github.com/mysiar/omnipay-przelewy24v1/pull/17)
- test-coverage-to-scrutinizer [`#15`](https://github.com/mysiar/omnipay-przelewy24v1/pull/15)
- Update issue templates [`#10`](https://github.com/mysiar/omnipay-przelewy24v1/pull/10)
- scrutinizer-ci-badge: update [`#14`](https://github.com/mysiar/omnipay-przelewy24v1/pull/14)
- cardInfo-fix [`f5831d7`](https://github.com/mysiar/omnipay-przelewy24v1/commit/f5831d7043543768c0ba06a5fa04787327eb3799)

#### [0.3.1](https://github.com/mysiar/omnipay-przelewy24v1/compare/0.3.0...0.3.1)

> 18 May 2022

- dead-code: update [`#12`](https://github.com/mysiar/omnipay-przelewy24v1/pull/12)
- main: update CHANGELOG [`98224a2`](https://github.com/mysiar/omnipay-przelewy24v1/commit/98224a2b768138e5f8306854143a242933a72427)

#### [0.3.0](https://github.com/mysiar/omnipay-przelewy24v1/compare/0.2.0...0.3.0)

> 18 May 2022

- ver 0.3.0 [`#11`](https://github.com/mysiar/omnipay-przelewy24v1/pull/11)
- [Refund] api/v1/transaction/refund [`#4`](https://github.com/mysiar/omnipay-przelewy24v1/pull/4)
- [Card Info] [`#9`](https://github.com/mysiar/omnipay-przelewy24v1/pull/9)
- main: update CHANGELOG [`0ff6152`](https://github.com/mysiar/omnipay-przelewy24v1/commit/0ff6152270c592d428f09ac878b36bfde73ee713)

#### [0.2.0](https://github.com/mysiar/omnipay-przelewy24v1/compare/0.1.0...0.2.0)

> 18 May 2022

- [Transaction by session] [`#8`](https://github.com/mysiar/omnipay-przelewy24v1/pull/8)
- transaction-register-more-params: updated PurchaseRequest with most of the P24 params [`#7`](https://github.com/mysiar/omnipay-przelewy24v1/pull/7)
- main: CHANGELOG [`ea0a7b7`](https://github.com/mysiar/omnipay-przelewy24v1/commit/ea0a7b7c1d7fbc00125ec529798b9b1ae34feb12)

#### 0.1.0

> 14 May 2022

- Transaction verify - completePurchase [`#6`](https://github.com/mysiar/omnipay-przelewy24v1/pull/6)
- missing-tests: PurchaseRequest test & refactor to use '10.55' values as amount [`#5`](https://github.com/mysiar/omnipay-przelewy24v1/pull/5)
- refactor-abstract-response: removed garbage endpoint property from AbstractResponse [`#3`](https://github.com/mysiar/omnipay-przelewy24v1/pull/3)
- Transaction register [`#2`](https://github.com/mysiar/omnipay-przelewy24v1/pull/2)
- methods: implemented [`#1`](https://github.com/mysiar/omnipay-przelewy24v1/pull/1)
- dev: TestAccess endpoint [`8404356`](https://github.com/mysiar/omnipay-przelewy24v1/commit/8404356e7a046be8adf15bb71b36d9c0740bde2b)
- dev: update README [`423d145`](https://github.com/mysiar/omnipay-przelewy24v1/commit/423d145ad29c6aad5d991ef7f385328d9a90f462)
- : init [`cb915fc`](https://github.com/mysiar/omnipay-przelewy24v1/commit/cb915fc2a23079ff1012d50b945bc1c70484de3d)


tests-api/GatewayTest.php
<?php

declare(strict_types=1);

namespace API;

use Omnipay\Omnipay;
use PHPUnit\Framework\TestCase;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\VarDumper\VarDumper;

class GatewayTest extends TestCase
{
    private $gateway;

    public function setUp(): void
    {
        $settings = [
            'merchantId' => getenv('P24V1_MERCHANT_ID'),
            'posId' => getenv('P24V1_POS_ID'),
            'crc' => getenv('P24V1_CRC'),
            'reportKey' => getenv('P24V1_REPORT_KEY'),
            'language' => getenv('P24V1_LANGUAGE'),
            'testMode' => true,
        ];

        $this->gateway = Omnipay::create('Przelewy24');
        $this->gateway->initialize($settings);
    }

    public function testAccess(): void
    {
        $response = $this->gateway->testAccess()->send();
        $this->assertSame(Response::HTTP_OK, $response->getCode());
        $this->assertSame('', $response->getMessage());
        $this->assertTrue($response->isSuccessful());

        $this->gateway->setReportKey('dummy');
        $response = $this->gateway->testAccess()->send();

        $this->assertFalse($response->isSuccessful());
        $this->assertSame(Response::HTTP_UNAUTHORIZED, $response->getCode());
        $this->assertSame('Incorrect authentication', $response->getMessage());
    }

    public function testMethods(): void
    {
        $response = $this->gateway->methods()->send();
        $this->assertSame(Response::HTTP_OK, $response->getCode());
        $this->assertSame('', $response->getMessage());
        $this->assertTrue($response->isSuccessful());
        $this->assertGreaterThan(0, count($response->getMethods()));
    }

    public function testPurchase(): void
    {
        $sessionId = $this->randomString();

        $response = $this->gateway->purchase([
            'sessionId' => $sessionId,
            'amount' => '10.66',
            'currency' => 'PLN',
            'description' => 'Transaction description',
            'email' => 'franek@dolas.com',
            'country' => 'PL',
            'returnUrl' => 'https://omnipay-przelewy24v1.requestcatcher.com/return',
            'notifyUrl' => 'https://omnipay-przelewy24v1.requestcatcher.com/notify',
        ])->send();

        VarDumper::dump($sessionId);
        VarDumper::dump($response->getRedirectUrl());
        VarDumper::dump($response->getMessage());

        $this->assertSame(Response::HTTP_OK, $response->getCode());
        $this->assertSame('', $response->getMessage());
        $this->assertTrue($response->isSuccessful());
        $this->assertContains('https://sandbox.przelewy24.pl/trnRequest/', $response->getRedirectUrl());
        $this->assertSame(35, strlen($response->getToken()));
    }

    /*
     * sequence to run this test
     * 1. run testPurchase
     * 2. use redirect url to commit transaction
     * 3. check request catcher for payload
     * 4. fill below sessionId & transactionId (orderId in payload) value
     */
    public function testComplete(): void
    {
        $this->markTestSkipped('Comment this out if you want complete purchase as per above comment description');
        $sessionId = 'u6ipAwPFXJiVTqG';
        $transactionId = '317213045';
        $response = $this->gateway->completePurchase([
            'sessionId' => $sessionId,
            'amount' => '10.66',
            'currency' => 'PLN',
            'transactionId' => $transactionId,
        ])->send();

        VarDumper::dump($response->getCode());
        VarDumper::dump($response->getMessage());

        $this->assertSame(Response::HTTP_OK, $response->getCode());
        $this->assertTrue($response->isSuccessful());
    }

    public function testPurchaseInfo(): void
    {
        $sessionId = '20c62d6b-5ff0-46a0-97eb-eea0dd5b4a93'; // real existing session
        $response = $this->gateway->purchaseInfo([
            'sessionId' => $sessionId,
        ])->send();

        VarDumper::dump($response->getCode());
        VarDumper::dump($response->getInfo());

        $this->assertSame(Response::HTTP_OK, $response->getCode());
        $this->assertTrue($response->isSuccessful());
        $this->assertCount(18, $response->getInfo());

        $response = $this->gateway->purchaseInfo('not-existing')->send();
        $this->assertSame(Response::HTTP_NOT_FOUND, $response->getCode());
    }

    public function testPurchaseCard(): void
    {
        $sessionId = $this->randomString();

        $request = $this->gateway->purchase([
            'sessionId' => $sessionId,
            'amount' => '10.66',
            'currency' => 'PLN',
            'description' => 'Transaction description - card',
            'email' => 'franek@dolas.com',
            'name' => 'Franek Dolas',
            'country' => 'PL',
            'returnUrl' => 'https://omnipay-przelewy24v1.requestcatcher.com/return',
            'notifyUrl' => 'https://omnipay-przelewy24v1.requestcatcher.com/notify',
            'cardNotifyUrl' => 'https://omnipay-przelewy24v1.requestcatcher.com/notifyCard',
            'channel' => 1,
        ]);

        $response = $request->send();

        VarDumper::dump($sessionId);
        VarDumper::dump($response->getRedirectUrl());
        VarDumper::dump($response->getMessage());

        $this->assertSame(Response::HTTP_OK, $response->getCode());
        $this->assertSame('', $response->getMessage());
        $this->assertTrue($response->isSuccessful());
        $this->assertContains('https://sandbox.przelewy24.pl/trnRequest/', $response->getRedirectUrl());
        $this->assertSame(35, strlen($response->getToken()));
    }

    public function testCardInfo(): void
    {
        $transactionId = '317229535';
        $this->markTestSkipped('Comment this out if you have payment with credit card');
        $response = $this->gateway->cardInfo([
            'transactionId' => $transactionId,
        ])->send();
        dump($response->getCode());
        dump($response->getInfo());
    }

    private function randomString(int $length = 15): string
    {
        $characters = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
        $charactersLength = strlen($characters);
        $randomString = '';
        for ($i = 0; $i < $length; ++$i) {
            $randomString .= $characters[random_int(0, $charactersLength - 1)];
        }

        return $randomString;
    }
}
